model/map-hot-rod/src/main/java/org/keycloak/models/map/storage/hotRod/locking/HotRodGlobalLockProvider.java

```diff
@@ -0,0 +1,87 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.models.map.storage.hotRod.locking;
+
+import org.infinispan.client.hotrod.MetadataValue;
+import org.infinispan.client.hotrod.RemoteCache;
+import org.jboss.logging.Logger;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.locking.GlobalLock;
+import org.keycloak.models.locking.GlobalLockProvider;
+import org.keycloak.models.locking.LockAcquiringTimeoutException;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.LinkedList;
+import java.util.List;
+
+
+public class HotRodGlobalLockProvider implements GlobalLockProvider {
+
+    private static final Logger LOG = Logger.getLogger(HotRodGlobalLockProvider.class);
+    private final RemoteCache<String, String> locksCache;
+    private final long defaultTimeoutMilliseconds;
+    private final List<GlobalLock> providedGlobalLocks = new LinkedList<>();
+
+    public HotRodGlobalLockProvider(KeycloakSession session, RemoteCache<String, String> locksCache, long defaultTimeoutMilliseconds) {
+        this.locksCache = locksCache;
+        this.defaultTimeoutMilliseconds = defaultTimeoutMilliseconds;
+
+        // Enlist this provider for closing so all provided locks are closed if not closed by AutoClosable after commit
+        session.enlistForClose(this);
+    }
+
+    @Override
+    public GlobalLock acquire(String lockName, Duration timeToWaitForLock) throws LockAcquiringTimeoutException {
+        if (timeToWaitForLock == null) {
+            // Set default timeout if null provided
+            timeToWaitForLock = Duration.ofMillis(defaultTimeoutMilliseconds);
+        }
+
+        LOG.debugf("Acquiring lock [%s].", lockName);
+        if (!HotRodLocksUtils.repeatPutIfAbsent(locksCache, lockName, timeToWaitForLock.toMillis(), 50)) {
+            MetadataValue<String> withMetadata = locksCache.getWithMetadata(lockName);
+            throw new LockAcquiringTimeoutException(lockName, withMetadata.getValue(), Instant.ofEpochMilli(withMetadata.getCreated()));
+        }
+        LOG.debugf("Lock acquired [%s].", lockName);
+        GlobalLock lock = new HotRodGlobalLock(() -> {
+            LOG.debugf("Releasing lock [%s].", lockName);
+            boolean result = HotRodLocksUtils.removeWithInstanceIdentifier(locksCache, lockName);
+            LOG.debugf("Lock [%s] release resulted with %s", lockName, result);
+
+            if (!result) {
+                LOG.errorf("Unlocking the lock [%s] was unsuccessful. Current lock value: [%s]", lockName, locksCache.get(lockName));
+                throw new IllegalStateException("Unlocking the lock was unsuccessful. See log for more details.");
+            }
+        });
+
+        // Keep reference to acquired lock, so it can be closed by on session close
+        providedGlobalLocks.add(lock);
```

Comment:
In the current state, removing the `lock` from the `providedGlobalLocks` would cause `ConcurrentModificationException` on line 83 where the list is traversed. Also, we can't use `this` in the lambda.

However, we can change this code to not use lambda and then make a snapshot of the list before traversing it. But what would be the advantage of it? Locks are currently checking whether they were already unlocked and do nothing if yes.

Reply:
The idea behind my comment was: I'd like the code to be symmetric - adding it when the lock is created, and removing it once it is released. When doing so, the `close()` method could log a warning when there are remaining entries in the list, as it indicates and API wasn't used correctly.

The `close()` might even throw an exception, so to force the developer to fix this.